pkg/gitops/deploykey/deploykey.go

```diff
@@ -0,0 +1,155 @@
+package deploykey
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/google/go-github/v31/github"
+	"github.com/kris-nova/logger"
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+	"github.com/weaveworks/eksctl/pkg/gitops/flux"
+	"golang.org/x/oauth2"
+)
+
+type Manager struct {
+	cluster  *api.ClusterMeta
+	repoURL  string
+	readOnly bool
+}
+
+func ForCluster(cluster *api.ClusterConfig) *Manager {
+	km := &Manager{
+		cluster: cluster.Metadata,
+	}
+
+	if git := cluster.Git; git != nil {
+		if repo := git.Repo; repo != nil {
+			km.repoURL = repo.URL
+		}
+
+		km.readOnly = git.ReadOnly
+	}
+
+	return km
+}
+
+func (km *Manager) Put(ctx context.Context, fluxSSHKey flux.PublicKey) error {
+	owner, repo, ok := km.getOwnerRepoFromRepoURL()
+	if !ok {
+		logger.Info("Skipped creating GitHub deploy key from Flux SSH public key for URL %s: Only `git@github.com:OWNER/REPO.git` is accepted for automatic deploy key creation", km.repoURL)
+
+		return nil
+	}
+
+	gh, ok := km.getGitHubAPIClient(ctx)
+	if !ok {
+		logger.Info("Skipped creating GitHub deploy key from Flux SSH public key due to missing GITHUB_TOKEN")
+
+		return nil
+	}
+
+	logger.Info("Creating GitHub deploy key from Flux SSH public key")
+
+	title := km.getDeployKeyTitle(km.cluster)
+
+	key, _, err := gh.Repositories.CreateKey(ctx, owner, repo, &github.Key{
+		Key:      &fluxSSHKey.Key,
+		Title:    &title,
+		ReadOnly: &km.readOnly,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	logger.Info("%s configured with Flux SSH public key\n%s", *key.Title, fluxSSHKey.Key)
+
+	return nil
+}
+
+func (km *Manager) Delete(ctx context.Context) error {
+	owner, repo, ok := km.getOwnerRepoFromRepoURL()
+	if !ok {
+		logger.Info("Skipped deleting GitHub deploy key for URL %s: Only `git@github.com:OWNER/REPO.git` is accepted for automatic deploy key creation", km.repoURL)
+
+		return nil
+	}
+
+	gh, ok := km.getGitHubAPIClient(ctx)
+	if !ok {
+		logger.Info("Skipped deleting GitHub deploy key due to missing GITHUB_TOKEN")
+
+		return nil
+	}
+
+	logger.Info("Deleting GitHub deploy key")
+
+	title := km.getDeployKeyTitle(km.cluster)
+
+	keys, _, err := gh.Repositories.ListKeys(ctx, owner, repo, &github.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	var keyID int64
+
+	for _, key := range keys {
+		if key.GetTitle() == title {
+			keyID = key.GetID()
+
+			break
+		}
+	}
+
+	if keyID == 0 {
+		logger.Info("Skipped deleting GitHub deploy key %q: The key does not exist. Probably you've already deleted it?")
+
+		return nil
+	}
+
+	if _, err := gh.Repositories.DeleteKey(ctx, owner, repo, keyID); err != nil {
+		return err
+	}
+
+	logger.Info("Deleted GitHub deploy key %s", title)
+
+	return nil
+}
+
+func (km *Manager) getGitHubAPIClient(ctx context.Context) (*github.Client, bool) {
+	githubToken := os.Getenv("GITHUB_TOKEN")
```

Comment:
Can you please create a constant for `GITHUB_TOKEN` and check its existence in `ForCluster`. If the env var is not set then all functions should be `no-op`. The `GitHubProvider` could have a `token` field with the value being set in `ForCluster`.